### PR TITLE
Add dgl

### PR DESCRIPTION
**Pull request by mikemhenry**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

There are a few issues with the recipe that I need to fix, I was trying to first get something working before I made it better:

~- [ ] Build for different OS, right now it is linux only~
~- [ ] Build a CPU version for people who don't want to pull in CUDA~
- [ ] Fix the vendored packages
- [x] Use official upstream force instead of my fork (I used that to get around needing to use git during the build phase and updated the submodules)

The crossed out TODOs will be done once we get this moved into a feedstock.

Supersedes https://github.com/conda-forge/staged-recipes/pull/16924
Thanks @knc6 for starting this work!

### CODE REVIEW

**Comment by hadim** [comment]
It's been a while and unsuccessful, but I also made an attempt at https://github.com/conda-forge/staged-recipes/pull/12552

File: recipes/dgl/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    - _openmp_mutex           # [linux]
+
+  run:
+    - python
+    - numpy
+    - scipy
+    - networkx
+    - requests
+    - tqdm
+    - pytorch =*={{ torch_proc_type }}*
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:43]:
```suggestion
    - pytorch * *{{ torch_proc_type }}*
```

File: recipes/dgl/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    - tensorflow
+    - _openmp_mutex           # [linux]
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:45]:
```suggestion
```

File: recipes/dgl/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}  # [cuda_compiler_version not in (undefined, 'None')]
+    - sysroot_linux-64  2.17  # [linux64]    
+    - cmake
+
+  host:
+    - python
+    - setuptools
+    - cython
+    - _openmp_mutex           # [linux]
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:34]:
```suggestion
```

File: recipes/dgl/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    - '*/libcuda.*'  # [linux64]
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}  # [cuda_compiler_version not in (undefined, 'None')]
+    - sysroot_linux-64  2.17  # [linux64]    
+    - cmake
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:28]:
```suggestion
    - cmake
    - llvm-openmp  # [osx]
    - libgomp      # [linux]
```

File: recipes/dgl/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    - scipy
+    - networkx
+    - requests
+    - tqdm
+    - pytorch =*={{ torch_proc_type }}*
+    - tensorflow
+    - _openmp_mutex           # [linux]
+
+  run_constrained:
+    - pytorch =*={{ torch_proc_type }}
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:48]:
```suggestion
```

**Comment by hadim** [comment]
Here is a small dgl python test script that could make the build even more robust. From the dgl doc:

```python
# Test from https://docs.dgl.ai/tutorials/models/1_gnn/1_gcn.html


import time
import numpy as np

import torch as th
import torch.nn as nn
import torch.nn.functional as F

import dgl
import dgl.function as fn
from dgl import DGLGraph
from dgl.data import CoraGraphDataset

gcn_msg = fn.copy_u(u="h", out="m")
gcn_reduce = fn.sum(msg="m", out="h")


class GCNLayer(nn.Module):
    def __init__(self, in_feats, out_feats):
        super(GCNLayer, self).__init__()
        self.linear = nn.Linear(in_feats, out_feats)

    def forward(self, g, feature):
        # Creating a local scope so that all the stored ndata and edata
        # (such as the `'h'` ndata below) are automatically popped out
        # when the scope exits.
        with g.local_scope():
            g.ndata["h"] = feature
            g.update_all(gcn_msg, gcn_reduce)
            h = g.ndata["h"]
            return self.linear(h)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.layer1 = GCNLayer(1433, 16)
        self.layer2 = GCNLayer(16, 7)

    def forward(self, g, features):
        x = F.relu(self.layer1(g, features))
        x = self.layer2(g, x)
        return x


def load_cora_data():
    dataset = CoraGraphDataset()
    g = dataset[0]
    features = g.ndata["feat"]
    labels = g.ndata["label"]
    train_mask = g.ndata["train_mask"]
    test_mask = g.ndata["test_mask"]
    return g, features, labels, train_mask, test_mask


def evaluate(model, g, features, labels, mask):
    model.eval()
    with th.no_grad():
        logits = model(g, features)
        logits = logits[mask]
        labels = labels[mask]
        _, indices = th.max(logits, dim=1)
        correct = th.sum(indices == labels)
        return correct.item() * 1.0 / len(labels)


def main():

    net = Net()

    g, features, labels, train_mask, test_mask = load_cora_data()
    # Add edges between each node and itself to preserve old node representations
    g.add_edges(g.nodes(), g.nodes())
    optimizer = th.optim.Adam(net.parameters(), lr=1e-2)
    dur = []
    t0 = time.time()
    for epoch in range(50):
        if epoch >= 3:
            t0 = time.time()

        net.train()
        logits = net(g, features)
        logp = F.log_softmax(logits, 1)
        loss = F.nll_loss(logp[train_mask], labels[train_mask])

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if epoch >= 3:
            dur.append(time.time() - t0)

        acc = evaluate(net, g, features, labels, test_mask)
        print(
            "Epoch {:05d} | Loss {:.4f} | Test Acc {:.4f} | Time(s) {:.4f}".format(
                epoch, loss.item(), acc, np.mean(dur)
            )
        )


if __name__ == "__main__":
    main()
```

**Comment by mikemhenry** [comment]
@hadim Thanks! This is a good idea, I will admit that I'm not much of a `dgl` user but I'm working to package this for our lab and others benefit. Will that code snippet work without a GPU? 

I've uploaded the cuda 10.2 build, the others failed because of diskspace issues, if you want to test it you can use `mamba install -c mmh dgl` to install it. I just tried the code snippet you posted and it ran without error!

**Comment by hadim** [comment]
Yes, the code is device-agnostic.

I don't know what is the recommended procedure here due to the disk space. Maybe simply merging as it is and see whether the feedstock CI pass on the failing cuda builds?

@jaimergp @conda-forge/core ?

**Comment by mikemhenry** [comment]
running out of diskspace is "normal" for these kinds of builds and will be fixed/fine when it is in a feedstock. 

There are a few issues with the recipe that I need to fix, I was trying to first get something working before I made it better:

- [ ] Build for different OS, right now it is linux only
- [ ] Build a CPU version for people who don't want to pull in CUDA
- [ ] Fix the vendored packages
- [ ] Use official upstream force instead of my fork (I used that to get around needing to use git during the build phase and updated the submodules)

So quite of work still to do, but this looks like it is all tractable! Do you know @hadim if we can get upstream on-board with this at all? I know they have their own conda package on `dglteam` but it isn't binary compatible with `conda-forge` and users clearly want this package on `conda-forge`. They know their cmake build system so they could probably allow it to find/use installed packages instead of those bundled more efficiently than I could patch it.

**Comment by jaimergp** [comment]
> but I think fixing vendored packages + using upstream should be done first IMHO.

Definitely, that's the priority right now, I'd say. Without that it's very unlikely this recipe can be recommended for merge.

CPU, non-Linux stuff can be worked on after that.

Can you move that check list in the description of the PR @mikemhenry? It'll make following the progress easier.

File: recipes/dgl/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+{% set version = "0.8.1" %}
+{% set torch_proc_type = "cuda" if cuda_compiler_version != "None" else "cpu" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  - git_url: https://github.com/dmlc/dgl
+  - git_rev: {{ version }}
```
Comment (jaimergp) [inline recipes/dgl/meta.yaml:11]:
```suggestion
  git_url: https://github.com/dmlc/dgl
  git_rev: {{ version }}
```

If you put it as a list it sees it as two sources, so I guess `git_rev` was being ignored!

**Comment by hadim** [comment]
I agree there is much to be done here (I haven't noticed you weren't building on for osx and win). I see you already commented on the ticket I opened a while ago @mikemhenry at https://github.com/dmlc/dgl/issues/1855.  

I will ping the dgl folks there again to see if they help here.

File: recipes/dgl/meta.yaml (hunk 6 above)
Comment (mikemhenry) [reply recipes/dgl/meta.yaml]: Thanks!

**Comment by mikemhenry** [comment]
> I agree there is more to be done here (I haven't noticed you weren't building for osx and win). I see you already commented on the ticket I opened a while ago @mikemhenry at [dmlc/dgl#1855](https://github.com/dmlc/dgl/issues/1855).
> 
> I will ping the dgl folks there again to see if they help here.

Thanks! Header only libraries are fine to vendor (we just need to make sure we package the license) but it would be great to have their help in modifying the build system to use installed packages (I will admit that I've only looked at the [CMakeLists.txt](https://github.com/dmlc/dgl/blob/master/CMakeLists.txt#L209) and it looks like there isn't an option to use installed libraries but I am not a CMake wizard)

**Comment by hadim** [comment]
Hey @mikemhenry, I am just checking whether you are still planning to work on that PR?

**Comment by mikemhenry** [comment]
@hadim I am -- but do you have any cmake experience? I've been busy with some other projects and haven't gotten to it, but I'm guessing it would be quick for someone who is proficient with cmake.

**Comment by hadim** [comment]
Thanks @mikemhenry. This is not urgent on my side, and I am quite busy as well, but I just wanted to know whether it was still on your radar or not. My cmake skills are quite old at this point to be honest xD

If you're too busy in the future, I might give it a try at some point but I really don't know when.

**Comment by mikemhenry** [comment]
@hadim made a wonderful table here: https://github.com/dmlc/dgl/issues/1855#issuecomment-1443916992 outlining which packages need to get onto conda-forge and which ones already exist. I am going to try and make the changes needed upstream, but I might use a patch if I get things working and upstream wants to take their time in evaluating it (which is their right :smile:  )

**Comment by mikemhenry** [comment]
we will get DGL on conda-forge here: https://github.com/conda-forge/staged-recipes/pull/22691

**Pull request closed without merging**